README.md
# notion-backup

A command line tool written in Haskell to download any Notion.so page as markdown.

## Usage

```
$ notion-backup --apiToken <your-api-token> --pageId ca0c8c0f-bfc3-4052-a178-35fbffccfce5
```


ChangeLog.md
# Changelog for notion-backup

## Unreleased changes


src/Utils.hs
{-# LANGUAGE OverloadedStrings #-}
{-# LANGUAGE StrictData #-}

module Utils
  ( buildNotionRequest,
    processResponse,
  )
where

import Control.Concurrent.MonadIO (MonadIO)
import Data.Aeson
import qualified Data.ByteString as B
import qualified Data.ByteString.Char8 as BC
import qualified Data.ByteString.Lazy as L
import qualified Data.ByteString.Lazy.Char8 as LC
import Data.Text as T
import Network.HTTP.Simple

apiHost :: BC.ByteString
apiHost = "www.notion.so"

headers :: BC.ByteString -> [Header]
headers notionToken =
  [ ("authority", "www.notion.so"),
    ("pragma", "no-cache"),
    ("cache-control", "no-cache"),
    ("content-type", "application/json"),
    ("user-agent", "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_5) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/84.0.4147.135 Safari/537.36"),
    ("notion-client-version", "23.2.10"),
    ("accept", "*/*"),
    ("origin", "https://www.notion.so"),
    ("sec-fetch-site", "same-origin"),
    ("sec-fetch-mode", "cors"),
    ("sec-fetch-dest", "empty"),
    ("referer", "https://www.notion.so/Test-one-block-ca0c8c0fbfc34052a17835fbffccfce5"),
    ("accept-language", "fr-FR,fr;q=0.9,en-US;q=0.8,en;q=0.7"),
    ("cookie", "token_v2=" <> notionToken)
  ]

buildRequest :: BC.ByteString -> BC.ByteString -> BC.ByteString -> BC.ByteString -> LC.ByteString -> Request
buildRequest token host method path requestBody =
  setRequestMethod method $
    setRequestHost host $
      setRequestHeaders (headers token) $
        setRequestPath path $
          setRequestSecure True $
            setRequestPort 443 $
              setRequestBodyLBS requestBody $
                defaultRequest

buildNotionRequest :: BC.ByteString -> BC.ByteString -> LC.ByteString -> Request
buildNotionRequest notionToken path body = buildRequest notionToken apiHost "POST" path body

processResponse :: Response L.ByteString -> Either String L.ByteString
processResponse response =
  if status == 200
    then Right jsonBody
    else Left $ "Error " <> statusText <> ": " <> show jsonBody
  where
    status = getResponseStatusCode response
    jsonBody = getResponseBody response
    statusText = show $ status


src/EnqueueTask.hs
{-# LANGUAGE OverloadedStrings #-}
{-# LANGUAGE StrictData #-}

module EnqueueTask
  ( requestEnqueueTask,
    taskIDEnqueueTaskResponse,
    decodeEnqueueTaskResponse,
  )
where

import Control.Concurrent.MonadIO (MonadIO)
import Data.Aeson
import qualified Data.ByteString as B
import qualified Data.ByteString.Char8 as BC
import qualified Data.ByteString.Lazy as L
import qualified Data.ByteString.Lazy.Char8 as LC
import Data.Text as T
import Network.HTTP.Simple
import Utils (buildNotionRequest)

enqueueTaskEndpoint :: BC.ByteString
enqueueTaskEndpoint = "/api/v3/enqueueTask"

data EnqueueTaskResponse = EnqueueTaskResponse
  { taskIDEnqueueTaskResponse :: Text
  }
  deriving (Show)

decodeEnqueueTaskResponse :: LC.ByteString -> Either String EnqueueTaskResponse
decodeEnqueueTaskResponse = eitherDecode

instance ToJSON EnqueueTaskResponse where
  toJSON (EnqueueTaskResponse taskIDEnqueueTaskResponse) =
    object
      [ "taskId" .= taskIDEnqueueTaskResponse
      ]

instance FromJSON EnqueueTaskResponse where
  parseJSON (Object v) =
    EnqueueTaskResponse
      <$> v .: "taskId"

-- Query body shape:
--
-- {
--   "task": {
--     "eventName": "exportBlock",
--     "request": {
--       "blockId": "00000000-0000-0000-0000-000000000000",
--       "recursive": false,
--       "exportOptions": {
--         "exportType": "markdown",
--         "timeZone": "Europe/Paris",
--         "locale": "en"
--       }
--     }
--   }
-- }
buildEnqueueTaskBody :: String -> LC.ByteString
buildEnqueueTaskBody pageBlockId =
  encode $
    object
      [ "task"
          .= object
            [ "eventName" .= ("exportBlock" :: String),
              "request"
                .= object
                  [ "blockId" .= (pageBlockId),
                    "recursive" .= (False),
                    "exportOptions"
                      .= object
                        [ "exportType" .= ("markdown" :: String),
                          "timeZone" .= ("Europe/Paris" :: String),
                          "locale" .= ("en" :: String)
                        ]
                  ]
            ]
      ]

requestEnqueueTask :: MonadIO m => String -> String -> m (Response L.ByteString)
requestEnqueueTask token pageBlockId = httpLBS request
  where
    request = buildNotionRequest (BC.pack token) enqueueTaskEndpoint $ buildEnqueueTaskBody pageBlockId


src/ExportPageCommand.hs
{-# LANGUAGE OverloadedStrings #-}
{-# LANGUAGE StrictData #-}

module ExportPageCommand (exportPageCommand) where

import Data.Aeson
import qualified Data.ByteString as B
import qualified Data.ByteString.Char8 as BC
import qualified Data.ByteString.Lazy as L
import qualified Data.ByteString.Lazy.Char8 as LC
import Data.Text as T
import EnqueueTask
  ( decodeEnqueueTaskResponse,
    requestEnqueueTask,
    taskIDEnqueueTaskResponse,
  )
import GetTasks
  ( getTasksRequest,
    getUrlFromTaskResponse,
  )
import Network.HTTP.Simple
import Utils
  ( buildNotionRequest,
    processResponse,
  )

processEnqueueTaskResponse :: Response L.ByteString -> Either String String
processEnqueueTaskResponse enqueueTaskResponse =
  processResponse enqueueTaskResponse
    >>= (\processedResponse -> decodeEnqueueTaskResponse processedResponse)
    >>= (\enqueueTaskResponse -> return $ T.unpack $ taskIDEnqueueTaskResponse enqueueTaskResponse)

download :: Text -> IO (Either String L.ByteString)
download url = do
  request <- parseRequest $ T.unpack url
  response <- httpLBS request
  return $ processResponse response

-- Enqueue task.
-- Deserializa json response.
-- Get task id.
-- Get tasks.
-- Deserialize json response.
-- Download from link.

-- TODO: Flatten this using monad transformers? See https://stackoverflow.com/questions/52016330/haskell-common-pattern-to-deal-with-failure-inside-io-io-either-string-int
-- TODO: Improve error messages.
exportPageCommand :: String -> String -> IO ()
exportPageCommand token pageId = do
  enqueueTaskResponse <- requestEnqueueTask token pageId
  case processEnqueueTaskResponse enqueueTaskResponse of
    Left errorMessage -> print $ "Error: " <> errorMessage
    Right taskId -> do
      print $ "taskId: " <> taskId
      getTasksResponse <- httpLBS $ getTasksRequest token taskId
      case getUrlFromTaskResponse getTasksResponse of
        Left errorMessage -> print $ "Error2: " <> errorMessage
        Right fileUrl -> do
          print $ "Downloading archive: " <> fileUrl
          eitherFileBytes <- download fileUrl
          case eitherFileBytes of
            Left errorMessage -> print $ "Error3: " <> errorMessage
            Right fileBytes -> L.writeFile "page.zip" fileBytes


src/GetTasks.hs
{-# LANGUAGE OverloadedStrings #-}
{-# LANGUAGE StrictData #-}

module GetTasks (getTasksRequest, getUrlFromTaskResponse) where

import Control.Concurrent.MonadIO (MonadIO)
import Data.Aeson
import qualified Data.ByteString as B
import qualified Data.ByteString.Char8 as BC
import qualified Data.ByteString.Lazy as L
import qualified Data.ByteString.Lazy.Char8 as LC
import Data.Text as T
import Network.HTTP.Simple
import Utils
  ( buildNotionRequest,
    processResponse,
  )

-- Response shape:
--
-- {
--   "results": [
--     {
--       "id": "00000000-0000-0000-0000-000000000000",
--       "eventName": "exportBlock",
--       "request": {
--         "blockId": "00000000-0000-0000-0000-000000000000",
--         "recursive": false,
--         "exportOptions": {
--           "exportType": "markdown",
--           "timeZone": "Europe/Paris",
--           "locale": "en"
--         }
--       },
--       "actor": {
--         "table": "notion_user",
--         "id": "00000000-0000-0000-0000-000000000000"
--       },
--       "state": "success",
--       "status": {
--         "type": "complete",
--         "pagesExported": 1,
--         "exportURL": "https://link.zip%22"
--       }
--     }
--   ]
-- }
data GetTaskResponse = GetTaskResponse
  { resultsGetTaskResponse :: [ResultElement]
  }
  deriving (Show)

data ResultElement = ResultElement
  { resultIDResultElement :: Text,
    eventNameResultElement :: Text,
    requestResultElement :: TaskRequest,
    actorResultElement :: Actor,
    stateResultElement :: Text,
    statusResultElement :: Status
  }
  deriving (Show)

data Actor = Actor
  { tableActor :: Text,
    actorIDActor :: Text
  }
  deriving (Show)

data TaskRequest = TaskRequest
  { blockIDRequest :: Text,
    recursiveRequest :: Bool,
    exportOptionsRequest :: ExportOptions
  }
  deriving (Show)

data ExportOptions = ExportOptions
  { exportTypeExportOptions :: Text,
    timeZoneExportOptions :: Text,
    localeExportOptions :: Text
  }
  deriving (Show)

data Status = Status
  { statusTypeStatus :: Text,
    pagesExportedStatus :: Int,
    exportURLStatus :: Text
  }
  deriving (Show)

decodeGetTaskResponse :: LC.ByteString -> Either String GetTaskResponse
decodeGetTaskResponse = eitherDecode

instance ToJSON GetTaskResponse where
  toJSON (GetTaskResponse resultsGetTaskResponse) =
    object
      [ "results" .= resultsGetTaskResponse
      ]

instance FromJSON GetTaskResponse where
  parseJSON (Object v) =
    GetTaskResponse
      <$> v .: "results"

instance ToJSON ResultElement where
  toJSON (ResultElement resultIDResultElement eventNameResultElement requestResultElement actorResultElement stateResultElement statusResultElement) =
    object
      [ "id" .= resultIDResultElement,
        "eventName" .= eventNameResultElement,
        "request" .= requestResultElement,
        "actor" .= actorResultElement,
        "state" .= stateResultElement,
        "status" .= statusResultElement
      ]

instance FromJSON ResultElement where
  parseJSON (Object v) =
    ResultElement
      <$> v .: "id"
      <*> v .: "eventName"
      <*> v .: "request"
      <*> v .: "actor"
      <*> v .: "state"
      <*> v .: "status"

instance ToJSON Actor where
  toJSON (Actor tableActor actorIDActor) =
    object
      [ "table" .= tableActor,
        "id" .= actorIDActor
      ]

instance FromJSON Actor where
  parseJSON (Object v) =
    Actor
      <$> v .: "table"
      <*> v .: "id"

instance ToJSON TaskRequest where
  toJSON (TaskRequest blockIDRequest recursiveRequest exportOptionsRequest) =
    object
      [ "blockId" .= blockIDRequest,
        "recursive" .= recursiveRequest,
        "exportOptions" .= exportOptionsRequest
      ]

instance FromJSON TaskRequest where
  parseJSON (Object v) =
    TaskRequest
      <$> v .: "blockId"
      <*> v .: "recursive"
      <*> v .: "exportOptions"

instance ToJSON ExportOptions where
  toJSON (ExportOptions exportTypeExportOptions timeZoneExportOptions localeExportOptions) =
    object
      [ "exportType" .= exportTypeExportOptions,
        "timeZone" .= timeZoneExportOptions,
        "locale" .= localeExportOptions
      ]

instance FromJSON ExportOptions where
  parseJSON (Object v) =
    ExportOptions
      <$> v .: "exportType"
      <*> v .: "timeZone"
      <*> v .: "locale"

instance ToJSON Status where
  toJSON (Status statusTypeStatus pagesExportedStatus exportURLStatus) =
    object
      [ "type" .= statusTypeStatus,
        "pagesExported" .= pagesExportedStatus,
        "exportURL" .= exportURLStatus
      ]

instance FromJSON Status where
  parseJSON (Object v) =
    Status
      <$> v .: "type"
      <*> v .: "pagesExported"
      <*> v .: "exportURL"

getTaskEndpoint :: BC.ByteString
getTaskEndpoint = "/api/v3/getTasks"

buildGetTasksBody :: String -> LC.ByteString
buildGetTasksBody taskId = "{\"taskIds\":[\"" <> LC.pack taskId <> "\"]}"

getTasksRequest :: String -> String -> Request
getTasksRequest token taskId = buildNotionRequest (BC.pack token) getTaskEndpoint $ buildGetTasksBody taskId

getExportUrl :: GetTaskResponse -> Text
getExportUrl getTaskResponse = exportURLStatus . statusResultElement . Prelude.head . resultsGetTaskResponse $ getTaskResponse

getUrlFromTaskResponse :: Response L.ByteString -> Either String Text
getUrlFromTaskResponse response =
  processResponse response
    >>= (\processedResponse -> decodeGetTaskResponse processedResponse)
    >>= \getTaskResponse -> return $ getExportUrl getTaskResponse
